testsuite/forge/src/success_criteria.rs

```diff
@@ -76,6 +89,98 @@ impl SuccessCriteria {
                 .await?;
         }
 
+        if let Some(chain_progress_threshold) = &self.chain_progress_check {
+            let epochs = FetchMetadata::fetch_new_block_events(
+                swarm.aptos_public_info().client(),
+                None,
+                None,
+            )
+            .await
+            .unwrap();
+
+            let mut max_round_gap = 0;
+            let mut max_round_gap_version = 0;
+            let mut max_time_gap = 0;
+            let mut max_time_gap_version = 0;
+
+            let mut prev_block = None;
+            let mut prev_ts = 0;
+            let mut failed_from_nil = 0;
+            let mut previous_epooch = 0;
+            let mut previous_round = 0;
+            for block in epochs
+                .iter()
+                .flat_map(|epoch| epoch.blocks.iter())
+                .filter(|b| b.version > start_version && b.version < end_version)
+            {
+                let is_nil = block.event.proposer() == PeerId::ZERO;
+
+                let current_gap = if previous_epooch == block.event.epoch() {
+                    block.event.round() - previous_round - 1
```

Comment:
Not sure I understand this - can you explain why we need this? the previous round will always be 1 less than the `block.event.round()` right? If that's the case, this is always 0?

Reply:
If rounds fail, there are no events for them. So this can be arbitrary large number

Reply:
Can you elaborate what you mean by round fail? IIUC, there should not be any gap in the round. The chain only moves forward one round if there is a QC or TC on that round.

Reply:
Proposer is defined for round number.  If proposer fails,  we go to the next round number, which defines a new proposer.

Height has no gaps. Not all rounds have NewBlockEvent

Reply:
Ah, I see - we don't have new block metadata even on nil block but the round moves forward. Makes sense.

Reply:
NIL block is different from TC.

NIL Block is 2f+1 validators agree on the same TC (consisting of same set of votes), which happens in small networks, but is highly unlikely in large. NIL blocks are blocks, and do have NewBlockEvent. But I don't want to treat them as successful here - so I count the gap across them. (which failed_from_nil captures)

We also have much more commonly a round timeouting with a TC, which doesn't generate a block or NewBlockEvent, and in that case NewBlockEvent is missing, what the formula here captures.